relay-server/src/services/project.rs

```diff
@@ -1126,21 +1126,51 @@ impl Project {
         outcome_aggregator: Addr<TrackOutcome>,
         buckets: Vec<Bucket>,
     ) -> Option<(Scoping, ProjectMetrics)> {
+        self.check_buckets_inner(outcome_aggregator, buckets)
+            .map_err(|e| {
+                match e {
+                    CheckBucketsError::ProjectExpired(len) => {
+                        relay_log::error!(
+                            tags.project_key = self.project_key.as_str(),
+                            "there is no project state: dropping {len} buckets"
+                        );
+                    }
+                    CheckBucketsError::ProjectDisabled(len) => {
+                        relay_log::debug!("dropping {len} buckets for disabled project");
+                    }
+
+                    CheckBucketsError::NoScoping(len) => {
+                        relay_log::error!(
+                            tags.project_key = self.project_key.as_str(),
+                            "there is no scoping: dropping {len} buckets"
+                        );
+                    }
+                    CheckBucketsError::RateLimited(len) => {
+                        relay_log::debug!("dropping {len} buckets due to rate limit");
+                    }
+                };
+                e
+            })
+            .ok()
+    }
+
+    fn check_buckets_inner(
+        &mut self,
+        outcome_aggregator: Addr<TrackOutcome>,
+        buckets: Vec<Bucket>,
+    ) -> Result<(Scoping, ProjectMetrics), CheckBucketsError> {
         let len = buckets.len();
+
         let Some(project_state) = self.valid_state() else {
-            relay_log::error!(
-                tags.project_key = self.project_key.as_str(),
-                "there is no project state: dropping {len} buckets"
-            );
-            return None;
+            return Err(CheckBucketsError::ProjectExpired(len));
         };
 
+        if project_state.invalid() || project_state.disabled() {
```

Comment:
There are helper functions like `metrics_allowed()` and `check_disabled()`, but they would be redundant here because they both do an expiry check, which in this case already happened on line 1164.